### PR TITLE
Fix Mikhail's bug

### PR DESCRIPTION
**Pull request by leerho**
This fixes Mikhail's bug along with a number of other related corner-case issues in the Set Operations.

### CODE REVIEW

**Comment by leerho** [comment]
An overview and explanation of the key changes contained in the latest updates can be read on the website:  
[https://datasketches.apache.org/docs/Theta/ThetaSetOpsCornerCases.html ](https://datasketches.apache.org/docs/Theta/ThetaSetOpsCornerCases.html )

File: src/main/java/org/apache/datasketches/tuple/CompactSketch.java [hunk 1]
```diff
@@ -118,9 +120,9 @@
         }
       }
     } else { // current serial format
-      offset++; // unused
+      offset++;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/CompactSketch.java:123]: will fix.  Thanks!

File: src/main/java/org/apache/datasketches/SetOperationCornerCases.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+package org.apache.datasketches;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SetOperationCornerCases {
+  private static final long MAX = Long.MAX_VALUE;
+
+  public enum InterResult {
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/SetOperationCornerCases.java:28]: will fix, thanks!

File: src/main/java/org/apache/datasketches/tuple/QuickSelectSketch.java [hunk 3]
```diff
@@ -51,7 +51,7 @@
   private int lgCurrentCapacity_;
   private final int lgResizeFactor_;
   private int count_;
-  private final SummaryFactory<S> summaryFactory_;
+  //private final SummaryFactory<S> summaryFactory_;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/QuickSelectSketch.java:54]: It is not needed here as it is already in the parent "Sketch"

File: src/main/java/org/apache/datasketches/SetOperationCornerCases.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      final long thetaLongB, final int countB, final boolean emptyB) {
+    return ((thetaLongA < MAX) ? 0 : 1 << 5)
+         | ((countA == 0)      ? 0 : 1 << 4)
+         | (!emptyA            ? 0 : 1 << 3)
+         | ((thetaLongB < MAX) ? 0 : 1 << 2)
+         | ((countB == 0)      ? 0 : 1 << 1)
+         | (!emptyB            ? 0 : 1);
+  }
+
+//  @Test
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/SetOperationCornerCases.java:187]: No, actually, it is a very useful when debugging, but I agree that it is in the wrong place. So I moved it to /tuple/MiscTest and leave a short reference to it above the relevant method here.  Thanks!

File: src/main/java/org/apache/datasketches/tuple/HashTables.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    count_ = sketch.getRetainedEntries(true);
+    lgTableSize_ = getLgTableSize(count_);
+    S mySummary = null;
+
+    hashTable_ = new long[1 << lgTableSize_];
+    final org.apache.datasketches.theta.HashIterator it = sketch.iterator();
+    while (it.next()) {
+      final long hash = it.get();
+      final int index = hashInsertOnly(hashTable_, lgTableSize_, hash);
+      mySummary = summary;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/HashTables.java]: You are correct.  Thanks!

File: src/main/java/org/apache/datasketches/tuple/HashTables.java [hunk 6]
```diff
@@ -80,13 +78,12 @@ void fromArrays(final long[] hashArr, final S[] summaryArr, final int count) {
...
     lgTableSize_ = getLgTableSize(count);
 
-    S mySummary = null;
     summaryTable_ = null;
     hashTable_ = new long[1 << lgTableSize_];
     for (int i = 0; i < count; i++) {
       final long hash = hashArr[i];
       final int index = hashInsertOnly(hashTable_, lgTableSize_, hash);
-      mySummary = summaryArr[i];
+      final S mySummary = summaryArr[i];
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/tuple/HashTables.java:86]: I suppose in this case it is fine not to make copies. The sketch can take ownership of the given objects, right?

File: src/main/java/org/apache/datasketches/SetOperationCornerCases.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    Exact_ResultDegen(060,      "A{ 1.0,>0, F} ; B{>1.0, 0, F}",
+        IntersectResult.RESULTDEGEN_MIN_0_F, AnotbResult.SKA_TRIM),   //48
+    Exact_NewDegen(061,         "A{ 1.0,>0, F} ; B{>1.0, 0, T}",
+        IntersectResult.NEW_1_0_T, AnotbResult.SKETCH_A),     //49
+    Exact_Estimation(062,       "A{ 1.0,>0, F} ; B{>1.0,>0, F}",
+        IntersectResult.FULL_INTER, AnotbResult.FULL_ANOTB),  //50
+    Exact_New(065,              "A{ 1.0,>0, F} ; B{ 1.0, 0, T}",
+        IntersectResult.NEW_1_0_T, AnotbResult.SKETCH_A),     //53
+    Exact_Exact(066,            "A{ 1.0,>0, F} ; B{ 1.0,>0, F}",
+        IntersectResult.FULL_INTER, AnotbResult.FULL_ANOTB);  //54
```
Comment (davecromberge) [inline src/main/java/org/apache/datasketches/SetOperationCornerCases.java:129]: Some of the descriptions for degenerative sketches use >1.0 for Theta which looks like a typo.

File: src/main/java/org/apache/datasketches/tuple/CompactSketch.java [hunk 8]
```diff
@@ -120,7 +120,7 @@
         }
       }
     } else { // current serial format
-      offset++;
+      offset++; //skip unused byte
       final byte flags = mem.getByte(offset++);
       offset += 2;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/CompactSketch.java:125]: This change is what you suggested in your previous review.  This is not "another", it is the same one :)

[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/tuple/CompactSketch.java:125]: I am referring to line 125, where was another "// unused" comment. I would suggest restoring it as well in a revised form: "// skip two unused bytes"

[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/tuple/CompactSketch.java:125]: or "skip unused bytes"

File: src/test/java/org/apache/datasketches/tuple/aninteger/MikhailsBugTupleTest.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    println(y.toString());
+    IntegerSummarySetOperations setOperations =
+        new IntegerSummarySetOperations(IntegerSummary.Mode.Min, IntegerSummary.Mode.Min);
+    Intersection<IntegerSummary> intersection = new Intersection<>(setOperations);
+    CompactSketch<IntegerSummary> intersect = intersection.intersect(x, y);
+    println("Tuple Intersect(Estimating, NewDegen) = new {1.0, 0, T}");
+    println(intersect.toString());
+    CompactSketch<IntegerSummary> csk = AnotB.aNotB(x, intersect);
+    println("Tuple AnotB(Estimating, New) = estimating {<1.0, 1, F}");
+    println(csk.toString());
```
Comment (davecromberge) [inline src/test/java/org/apache/datasketches/tuple/aninteger/MikhailsBugTupleTest.java:64]: Would it be a good idea to use SetOperationCornerCases to verify the corner case matches - or is this unnecessary?

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+  private static final double MIDP_THETA = MIDP;
+  private static final double LOWP_THETA = LOWP;
+
+  private enum SkType {
+    NEW,          //{ 1.0,  0, T} Bin: 101  Oct: 05
+    EXACT,        //{ 1.0, >0, F} Bin: 111  Oct: 07, specify only value
+    ESTIMATION,   //{<1.0, >0, F} Bin: 010  Oct: 02, specify only value
+    NEW_DEGEN,    //{<1.0,  0, T} Bin: 001  Oct: 01, specify only p
+    RESULT_DEGEN  //{<1.0,  0, F} Bin: 000  Oct: 0, specify p, value
+  }
```
Comment (davecromberge) [inline src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:57]: The website assigns 6 as octal id for the EXACT sketch type, which is incorrect - I have created a PR to update this on the website docs.

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+    final double resultInterTheta = 1.0;
+    final int resultInterCount = 0;
+    final boolean resultInterEmpty = true;
+    final double resultAnotbTheta = 1.0;
+    final int resultAnotbCount = 0;
+    final boolean resultAnotbEmpty = true;
+
+    checks(thetaA, thetaB, resultInterTheta, resultInterCount, resultInterEmpty,
+        resultAnotbTheta, resultAnotbCount, resultAnotbEmpty);
+  }
```
Comment (davecromberge) [inline src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:128]: What effect does a sampling probability of zero have on whether a sketch stores a value in its buffer?  This question is in reference to line 118 where an exact sketch is created with a value greater than the MIDP_V.

**Review by davecromberge — APPROVED** [review]

**Comment by davecromberge** [comment]
The related documentation on the website is very instructive and helped with my previous confusion about the different cases that were being tested.   The use of enums to convey the corner case in question back to the AnotB set operation was also a nice approach.

File: src/main/java/org/apache/datasketches/tuple/HashTables.java (hunk 6 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/HashTables.java:86]: In this case, it is fine.  

There are 2 ways to initialize the HashTables class: `fromSketch(...tuple.Sketch<S>)` and `fromSketch(...theta.Sketch)`.  In both cases the elements of the internal summaries arrays are copies of the summaries of the presented sketch, or copies of the provided summary in the case of theta.

The method you ask about, `fromArrays(...)`, is a private method (now private :) ) and only called by the 2 `getIntersectHashTables(next*Sketch,...)` methods.   These two methods rely on the `SummarySetOperations<S>` class to perform the configured set operation with the internal state of `HashTables` and the provided `next*Sketch`.   

Meanwhile, the Javadoc of the `SummarySetOperations<S>` interface has a clear warning for implementers:

```
Caution: Do not modify the input Summary objects. Also do not return them directly,
   unless they are immutable (most Summary objects are not). For mutable Summary objects, it is
   important to create a new Summary object with the correct contents to be returned. Do not
   return null summaries.
```
Assuming that this interface is implemented correctly by the user, there should not be any Summary objects belonging to input sketches that end up in the internals of these set operations.  This means that the extra `copy()`, line 150, I added in the last review should not be necessary.

A further question is whether we need to make sure no internal summary objects end up in result sketches.  I'm going to have to think more about this, but I think we may be OK, as the internals of these set operations are frequently cleared, ... nonetheless :)

File: src/main/java/org/apache/datasketches/SetOperationCornerCases.java (hunk 7 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/SetOperationCornerCases.java:129]: Thank you!

File: src/main/java/org/apache/datasketches/tuple/CompactSketch.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/CompactSketch.java:125]: Agreed.  Sorry I missed it.

File: src/main/java/org/apache/datasketches/tuple/AnotB.java [hunk 12]
```diff
@@ -177,23 +215,61 @@ public void notB(final Sketch<S> skB) {
...
+        hashArr_ = da.hashArr;
+        curCount_ = (hashArr_ == null) ? 0 : hashArr_.length;
+        summaryArr_ = da.summaryArr;
+        break;
+      }
+      case SKETCH_A: {
+        break; //result is already in A
+      }
+      case FULL_ANOTB: { //both A and B should have valid entries.
+        thetaLong_ = min(thetaLong_, skB.getThetaLong());
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/AnotB.java:260]: You are correct!

File: src/main/java/org/apache/datasketches/tuple/AnotB.java [hunk 13]
```diff
@@ -143,19 +142,58 @@ public void setA(final Sketch<S> skA) {
...
+
+    final int id =
+        SetOperationCornerCases.createCornerCaseId(thetaLong_, curCount_, empty_, thetaLongB, countB, emptyB);
+    final CornerCase cCase = CornerCase.idToCornerCase(id);
+    final AnotbResult anotbResult = cCase.getAnotbResult();
+
+    switch (anotbResult) {
+      case NEW_1_0_T: {
+        reset();
+        break;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/AnotB.java:161]: Ah, an excellent observation and good question!  The New Degenerate {<1.0, 0, T} is a special temporary state in that it has never seen any data.  When the first item is presented to a New Degenerate sketch the T becomes F, and the count will only be changed to 1 if the hash of the item is <theta, which becomes {<1.0, 1, F}, otherwise it becomes {<1.0, 0, F}; both of which are stable states.

Because the New Degenerate sketch has never seen any data, it is interpreted as if it were just New {1.0, 0, T} in any set operation.

File: src/main/java/org/apache/datasketches/tuple/AnotB.java [hunk 14]
```diff
@@ -143,19 +142,58 @@ public void setA(final Sketch<S> skA) {
...
+        reset();
+        thetaLong_ = min(thetaLong_, thetaLongB);
+        empty_ = false;
+        break;
+      }
+      case RESULTDEGEN_THA_0_F: {
+        empty_ = false;
+        curCount_ = 0;
+        //thetaLong_ is ok
+        break;
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/AnotB.java:173]: Yes!  I thought about writing out the math for this for the website, but it was too complicated for this particular subject page.  It really belongs in a more math oriented document.

File: src/main/java/org/apache/datasketches/tuple/CompactSketch.java [hunk 15]
```diff
@@ -240,6 +242,34 @@ public int getCountLessThanThetaLong(final long thetaLong) {
...
+      }
+    }
+
+    qsSk.setThetaLong(thetaLong);
+    if (countOut == 0) {
+      if (thetaLong == Long.MAX_VALUE) {
+        return new CompactSketch<>(null, null, thetaLong, true);
+      } else {
+        return new CompactSketch<>(null, null, thetaLong, false);
+      }
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/tuple/CompactSketch.java:267]: The containing method was a dead method.  Nothing calls it.  This was left over from the previous set operation logic. I probably have other dead code, which I will look for. Thanks!

File: src/test/java/org/apache/datasketches/tuple/aninteger/MikhailsBugTupleTest.java (hunk 9 above)
Comment (leerho) [reply src/test/java/org/apache/datasketches/tuple/aninteger/MikhailsBugTupleTest.java:64]: This code is a replica of Mikhail's code as he wrote it, plus some print statements.  I make it a practice to place test code of discovered bugs, among the tests as a way of making sure we covered the case he found.  Nonetheless, the tests that I constructed are far more comprehensive and covers his case and many others.

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java (hunk 10 above)
Comment (leerho) [reply src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:57]: No. The website is correct, the comment next to (now line 63 in this class) was incorrectly marked as 7.  It should be 6.  But in this test class, these octal values are not used.

Comment (leerho) [reply src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:57]: Oops the line number is 53 (not 63). But my comment still stands.

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java (hunk 11 above)
Comment (leerho) [reply src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:128]: If you look at line 53 (corrected) it reads:
`EXACT,        //{ 1.0, >0, F} Bin: 110  Oct: 06, specify only value`

And line 59:
`//NOTE: 0 values in getSketch are not used.`
Examine lines 512-515, p is not used.

So for line 118:
`UpdateSketch thetaB = getSketch(SkType.EXACT,  0, GT_MIDP_V);`

I could have used any integer value for this test, I just happened to choose GT_MIDP_V (3).  The 0 is not used.

**Comment by leerho** [comment]
If there are no further issues, I will merge this PR as it more than satisfies the objective of fixing Mikhail's bug.  Nonetheless, there are some other checks that I would like to do, but those can be in separate PRs.  I'll leave this open for another 24 hours.

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java (hunk 10 above)
Comment (davecromberge) [reply src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:57]: I incorrectly interpreted the bit for empty, thanks for clarifying.

File: src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java (hunk 11 above)
Comment (davecromberge) [reply src/test/java/org/apache/datasketches/theta/CornerCaseThetaSetOperationsTest.java:128]: Makes sense, thanks.

File: src/main/java/org/apache/datasketches/tuple/AnotB.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (davecromberge) [reply src/main/java/org/apache/datasketches/tuple/AnotB.java:161]: Thanks for the additional details, I understand the reason now.

**Review by AlexanderSaydakov — APPROVED** [review]

**Comment by AlexanderSaydakov** [comment]
I expect to take a closer look at the logic while porting the tests (and, perhaps, adjusting the implementation) to C++
